controllers/machine_controller_node.go

```diff
@@ -76,13 +77,50 @@ func (r *MachineReconciler) reconcileNodeRef(ctx context.Context, cluster *clust
 	}
 
 	// Set the Machine NodeRef.
-	machine.Status.NodeRef = nodeRef
-	logger.Info("Set Machine's NodeRef", "noderef", machine.Status.NodeRef.Name)
-	r.recorder.Event(machine, apicorev1.EventTypeNormal, "SuccessfulSetNodeRef", machine.Status.NodeRef.Name)
+	if machine.Status.NodeRef == nil {
+		machine.Status.NodeRef = &apicorev1.ObjectReference{
+			Kind:       node.Kind,
+			APIVersion: node.APIVersion,
+			Name:       node.Name,
+			UID:        node.UID,
+		}
+		logger.Info("Set Machine's NodeRef", "noderef", machine.Status.NodeRef.Name)
+		r.recorder.Event(machine, apicorev1.EventTypeNormal, "SuccessfulSetNodeRef", machine.Status.NodeRef.Name)
+		conditions.MarkFalse(machine, clusterv1.MachineNodeHealthyCondition, clusterv1.NodeProvisioningReason, clusterv1.ConditionSeverityWarning, "")
+		return nil
+	}
+
+	// Do the remaining node health checks, then set the node health to true if all checks pass.
+	if !checkNodeConditions(node) {
+		conditions.MarkFalse(machine, clusterv1.MachineNodeHealthyCondition, clusterv1.NodeConditionsFailedReason, clusterv1.ConditionSeverityWarning, "")
+		return nil
+	}
+	conditions.MarkTrue(machine, clusterv1.MachineNodeHealthyCondition)
+
 	return nil
 }
 
-func (r *MachineReconciler) getNodeReference(c client.Reader, providerID *noderefutil.ProviderID) (*apicorev1.ObjectReference, error) {
+// checkNodeConditions checks if a Node is healthy and does not have any semantically negative Conditions.
+// Returns true if NodeMemoryPressure, NodeDiskPressure, or NodePIDPressure Conditions are false or Ready Condition is true.
+func checkNodeConditions(node *apicorev1.Node) bool {
+	for _, condition := range node.Status.Conditions {
+		if condition.Type == apicorev1.NodeMemoryPressure && condition.Status == apicorev1.ConditionTrue {
+			return false
+		}
+		if condition.Type == apicorev1.NodeDiskPressure && condition.Status == apicorev1.ConditionTrue {
+			return false
+		}
+		if condition.Type == apicorev1.NodePIDPressure && condition.Status == apicorev1.ConditionTrue {
+			return false
+		}
+		if condition.Type == apicorev1.NodeReady && condition.Status == apicorev1.ConditionFalse {
+			return false
+		}
```

Comment:
Can we have a switch case here instead of listing if conditions one by one?

As an example:
```suggestion
		switch condition.Type {
			case apicorev1.NodeMemoryPressure, NodeDiskPressure, NodePIDPressure:
				return condition.Status != apicorev1.ConditionTrue
			[...] and so on [...]
		}
```